File: pm_core/src/infrastructure/store/adapter/priority_scheme_adapter.go
package adapter

import (
	"context"
	"fmt"

	"github.com/serp/pm-core/src/core/domain/entity"
	"github.com/serp/pm-core/src/core/port/store"
	"github.com/serp/pm-core/src/infrastructure/store/mapper"
	"github.com/serp/pm-core/src/infrastructure/store/model"
	"gorm.io/gorm"
)

type PrioritySchemeAdapter struct {
	db         *gorm.DB
	mapper     *mapper.PrioritySchemeMapper
	itemMapper *mapper.PrioritySchemeItemMapper
}

func NewPrioritySchemeAdapter(db *gorm.DB) store.IPrioritySchemeStore {
	return &PrioritySchemeAdapter{
		db:         db,
		mapper:     mapper.NewPrioritySchemeMapper(),
		itemMapper: &mapper.PrioritySchemeItemMapper{},
	}
}

func (a *PrioritySchemeAdapter) CreatePriorityScheme(ctx context.Context, tx *gorm.DB, scheme *entity.PrioritySchemeEntity) (*entity.PrioritySchemeEntity, error) {
	db := a.getDB(tx)
	schemeModel := a.mapper.ToModel(scheme)
	if err := db.WithContext(ctx).Create(schemeModel).Error; err != nil {
		return nil, fmt.Errorf("failed to create priority scheme: %w", err)
	}
	return a.mapper.ToEntity(schemeModel), nil
}

func (a *PrioritySchemeAdapter) UpdatePriorityScheme(ctx context.Context, tx *gorm.DB, scheme *entity.PrioritySchemeEntity) error {
	db := a.getDB(tx)
	schemeModel := a.mapper.ToModel(scheme)
	if err := db.WithContext(ctx).Save(schemeModel).Error; err != nil {
		return fmt.Errorf("failed to update priority scheme: %w", err)
	}
	return nil
}

func (a *PrioritySchemeAdapter) SoftDeletePriorityScheme(ctx context.Context, tx *gorm.DB, schemeID, tenantID int64) error {
	db := a.getDB(tx)
	if err := db.WithContext(ctx).
		Model(&model.PrioritySchemeModel{}).
		Where("id = ? AND tenant_id = ? AND deleted_at IS NULL", schemeID, tenantID).
		Update("deleted_at", gorm.Expr("CURRENT_TIMESTAMP")).Error; err != nil {
		return fmt.Errorf("failed to soft delete priority scheme: %w", err)
	}
	return nil
}

func (a *PrioritySchemeAdapter) GetPrioritySchemeByID(ctx context.Context, schemeID, tenantID int64) (*entity.PrioritySchemeEntity, error) {
	var schemeModel model.PrioritySchemeModel
	if err := a.db.WithContext(ctx).
		Where("id = ? AND tenant_id = ? AND deleted_at IS NULL", schemeID, tenantID).
		First(&schemeModel).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil
		}
		return nil, fmt.Errorf("failed to get priority scheme by id: %w", err)
	}
	return a.mapper.ToEntity(&schemeModel), nil
}

func (a *PrioritySchemeAdapter) GetPrioritySchemeWithItems(ctx context.Context, schemeID, tenantID int64) (*entity.PrioritySchemeEntity, error) {
	schemeEntity, err := a.GetPrioritySchemeByID(ctx, schemeID, tenantID)
	if err != nil || schemeEntity == nil {
		return schemeEntity, err
	}

	var itemModels []*model.PrioritySchemeItemModel
	if err := a.db.WithContext(ctx).
		Where("scheme_id = ? AND tenant_id = ? AND deleted_at IS NULL", schemeID, tenantID).
		Order("sequence ASC").
		Find(&itemModels).Error; err != nil {
		return nil, fmt.Errorf("failed to get priority scheme items: %w", err)
	}

	for _, itemModel := range itemModels {
		schemeEntity.Items = append(schemeEntity.Items, *a.itemMapper.ToEntity(itemModel))
	}
	return schemeEntity, nil
}

func (a *PrioritySchemeAdapter) ListPrioritySchemes(ctx context.Context, tenantID int64) ([]*entity.PrioritySchemeEntity, error) {
	var schemeModels []*model.PrioritySchemeModel
	if err := a.db.WithContext(ctx).
		Where("tenant_id = ? AND deleted_at IS NULL", tenantID).
		Order("created_at DESC").
		Find(&schemeModels).Error; err != nil {
		return nil, fmt.Errorf("failed to list priority schemes: %w", err)
	}
	return a.mapper.ToEntities(schemeModels), nil
}

func (a *PrioritySchemeAdapter) ExistsByName(ctx context.Context, tenantID int64, name string) (bool, error) {
	var count int64
	if err := a.db.WithContext(ctx).Model(&model.PrioritySchemeModel{}).
		Where("tenant_id = ? AND name = ? AND deleted_at IS NULL", tenantID, name).
		Count(&count).Error; err != nil {
		return false, fmt.Errorf("failed to check priority scheme name existence: %w", err)
	}
	return count > 0, nil
}

func (a *PrioritySchemeAdapter) getDB(tx *gorm.DB) *gorm.DB {
	if tx != nil {
		return tx
	}
	return a.db
}

File: pm_core/src/main/java/serp/project/pmcore/infrastructure/store/adapter/PrioritySchemeAdapter.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.pmcore.infrastructure.store.adapter;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Component;

import serp.project.pmcore.domain.priority.entity.PrioritySchemeEntity;
import serp.project.pmcore.domain.priority.port.IPrioritySchemePort;
import serp.project.pmcore.domain.priority.query.PrioritySchemeListCriteria;
import serp.project.pmcore.domain.shared.pagination.PageResult;
import serp.project.pmcore.infrastructure.store.mapper.PrioritySchemeItemMapper;
import serp.project.pmcore.infrastructure.store.mapper.PrioritySchemeMapper;
import serp.project.pmcore.infrastructure.store.model.PrioritySchemeModel;
import serp.project.pmcore.infrastructure.store.repository.IPrioritySchemeItemRepository;
import serp.project.pmcore.infrastructure.store.repository.IPrioritySchemeRepository;
import serp.project.pmcore.infrastructure.store.support.PageableUtils;

import java.util.List;
import java.util.Optional;

@Component
@RequiredArgsConstructor
public class PrioritySchemeAdapter implements IPrioritySchemePort {

    private final IPrioritySchemeRepository prioritySchemeRepository;
    private final IPrioritySchemeItemRepository prioritySchemeItemRepository;
    private final PrioritySchemeMapper prioritySchemeMapper;
    private final PrioritySchemeItemMapper prioritySchemeItemMapper;

    @Override
    public PrioritySchemeEntity createPriorityScheme(PrioritySchemeEntity scheme) {
        return prioritySchemeMapper.toEntity(prioritySchemeRepository.save(prioritySchemeMapper.toModel(scheme)));
    }

    @Override
    public Optional<PrioritySchemeEntity> getPrioritySchemeById(Long schemeId, Long tenantId) {
        return prioritySchemeRepository.findByIdAndTenantId(schemeId, tenantId)
                .map(prioritySchemeMapper::toEntity);
    }

    @Override
    public Optional<PrioritySchemeEntity> getPrioritySchemeByIdIncludingSystem(Long schemeId, Long tenantId) {
        return prioritySchemeRepository.findByIdAndTenantIdOrSystemTenant(schemeId, tenantId)
                .map(prioritySchemeMapper::toEntity);
    }

    @Override
    public Optional<PrioritySchemeEntity> getPrioritySchemeWithItems(Long schemeId, Long tenantId) {
        return prioritySchemeRepository.findByIdAndTenantId(schemeId, tenantId)
                .map(model -> {
                    PrioritySchemeEntity scheme = prioritySchemeMapper.toEntity(model);
                    scheme.setItems(prioritySchemeItemMapper.toEntities(
                            prioritySchemeItemRepository.findAllByTenantIdAndSchemeIdOrderBySequenceAsc(tenantId, schemeId)
                    ));
                    return scheme;
                });
    }

    @Override
    public List<PrioritySchemeEntity> listPrioritySchemes(Long tenantId) {
        return prioritySchemeMapper.toEntities(prioritySchemeRepository.findAllByTenantIdOrderByCreatedAtDesc(tenantId));
    }

    @Override
    public PageResult<PrioritySchemeEntity> listPrioritySchemesIncludingSystem(Long tenantId,
                                                                               PrioritySchemeListCriteria criteria) {
        Pageable pageable = PageableUtils.of(criteria, resolveSort(criteria));
        Page<PrioritySchemeModel> result = prioritySchemeRepository.findAllVisibleWithFilters(
                tenantId,
                criteria.getSearch(),
                criteria.getIsSystem(),
                pageable
        );
        return new PageResult<>(prioritySchemeMapper.toEntities(result.getContent()), result.getTotalElements());
    }

    @Override
    public void updatePriorityScheme(PrioritySchemeEntity scheme) {
        prioritySchemeRepository.save(prioritySchemeMapper.toModel(scheme));
    }

    @Override
    public void deletePriorityScheme(Long schemeId, Long tenantId) {
        prioritySchemeRepository.deleteByIdAndTenantId(schemeId, tenantId);
    }

    @Override
    public boolean existsByName(Long tenantId, String name) {
        return prioritySchemeRepository.existsByTenantIdAndName(tenantId, name);
    }

    @Override
    public boolean existsByDefaultPriorityId(Long priorityId, Long tenantId) {
        return prioritySchemeRepository.existsByDefaultPriorityIdAndTenantId(priorityId, tenantId);
    }

    private Sort resolveSort(PrioritySchemeListCriteria criteria) {
        String sortBy = criteria.getSortBy().toLowerCase();
        Sort.Direction direction = PageableUtils.resolveDirection(criteria.getSortDirection());

        return switch (sortBy) {
            case "name" -> Sort.by(
                    new Sort.Order(direction, "name"),
                    new Sort.Order(direction, "id")
            );
            case "created_at" -> Sort.by(
                    new Sort.Order(direction, "createdAt"),
                    new Sort.Order(direction, "id")
            );
            default -> throw new IllegalArgumentException("sortBy must be one of name, created_at");
        };
    }
}
